### PR TITLE
UPLOAD-1648/redis-client-improvements

### PR DESCRIPTION
**Pull request by cfarmer-fearless**
This patch is to address some recent issues found with our redis file locker implementation.  Some highlights include
- Better default logic so the server loudly fails to start up if a redis URL is provided but fails to connect and init the client.
- Additional nil checks when using the client.  The original outage was caused by a panic from referencing the client, which became nil at some point.  It is still not clear what caused it to become nil.
- Building off the last point, we may want to create a new client for every upload instead of having a long lived client.

Stack trace from panic that caused prd outage:
```
2024-08-12T09:33:35.384457455Z 2024/08/12 09:33:35 http: panic serving 169.254.129.1:10955: runtime error: invalid memory address or nil pointer dereference

2024-08-12T09:33:35.384481156Z goroutine 85790326 [running]:

2024-08-12T09:33:35.384487657Z net/http.(*conn).serve.func1()

2024-08-12T09:33:35.384492157Z 	/usr/local/go/src/net/http/server.go:1898 +0xbe

2024-08-12T09:33:35.384496457Z panic({0x94db60?, 0xe75f10?})

2024-08-12T09:33:35.384501257Z 	/usr/local/go/src/runtime/panic.go:770 +0x132

2024-08-12T09:33:35.384505758Z github.com/cdcgov/data-exchange-upload/upload-server/pkg/redislocker.(*RedisLocker).NewLock(0x0, {0xc000844c0c, 0x20})

2024-08-12T09:33:35.384510058Z 	/app/pkg/redislocker/redislocker.go:117 +0x79

2024-08-12T09:33:35.384513958Z github.com/tus/tusd/v2/pkg/handler.(*UnroutedHandler).lockUpload(0xc0001b2600, 0xc0005c0a50, {0xc000844c0c?, 0x2?})

2024-08-12T09:33:35.384517958Z 	/go/pkg/mod/github.com/tus/tusd/v2@v2.4.0/pkg/handler/unrouted_handler.go:1338 +0x5d

2024-08-12T09:33:35.384522259Z github.com/tus/tusd/v2/pkg/handler.(*UnroutedHandler).HeadFile(0xc0001b2600, {0xafa258?, 0xc0006fd0a0?}, 0xc00024b8c0)

2024-08-12T09:33:35.384527459Z 	/go/pkg/mod/github.com/tus/tusd/v2@v2.4.0/pkg/handler/unrouted_handler.go:621 +0x1b2

2024-08-12T09:33:35.384530459Z github.com/tus/tusd/v2/pkg/handler.NewHandler.func1({0xafa258, 0xc0006fd0a0}, 0xc00024b8c0)

2024-08-12T09:33:35.384533159Z 	/go/pkg/mod/github.com/tus/tusd/v2@v2.4.0/pkg/handler/handler.go:55 +0x251

2024-08-12T09:33:35.384535759Z net/http.HandlerFunc.ServeHTTP(0xc000fbb320?, {0xafa258?, 0xc0006fd0a0?}, 0xa1fa45?)

2024-08-12T09:33:35.384538459Z 	/usr/local/go/src/net/http/server.go:2166 +0x29

2024-08-12T09:33:35.384540859Z github.com/tus/tusd/v2/pkg/handler.NewHandler.(*UnroutedHandler).Middleware.func2({0xafa258, 0xc0006fd0a0}, 0xc00024b560)

2024-08-12T09:33:35.384543260Z 	/go/pkg/mod/github.com/tus/tusd/v2@v2.4.0/pkg/handler/unrouted_handler.go:267 +0x78c

2024-08-12T09:33:35.384545660Z net/http.HandlerFunc.ServeHTTP(0x40eba5?, {0xafa258?, 0xc0006fd0a0?}, 0xc0001e1a01?)

2024-08-12T09:33:35.384548260Z 	/usr/local/go/src/net/http/server.go:2166 +0x29

2024-08-12T09:33:35.384570661Z github.com/cdcgov/data-exchange-upload/upload-server/cmd/cli.Serve.StripPrefix.func3({0xafa258, 0xc0006fd0a0}, 0xc00024b440)

2024-08-12T09:33:35.384573461Z 	/usr/local/go/src/net/http/server.go:2209 +0x262

2024-08-12T09:33:35.384575961Z net/http.HandlerFunc.ServeHTTP(0xe90380?, {0xafa258?, 0xc0006fd0a0?}, 0x68a09a?)

2024-08-12T09:33:35.384578461Z 	/usr/local/go/src/net/http/server.go:2166 +0x29

2024-08-12T09:33:35.384580861Z net/http.(*ServeMux).ServeHTTP(0x468719?, {0xafa258, 0xc0006fd0a0}, 0xc00024b440)

2024-08-12T09:33:35.384583362Z 	/usr/local/go/src/net/http/server.go:2683 +0x1ad

2024-08-12T09:33:35.384586262Z net/http.serverHandler.ServeHTTP({0xc000fbb170?}, {0xafa258?, 0xc0006fd0a0?}, 0x6?)

2024-08-12T09:33:35.384588762Z 	/usr/local/go/src/net/http/server.go:3137 +0x8e

2024-08-12T09:33:35.384591062Z net/http.(*conn).serve(0xc0004a0c60, {0xaface8, 0xc000388240})

2024-08-12T09:33:35.384593762Z 	/usr/local/go/src/net/http/server.go:2039 +0x5e8

2024-08-12T09:33:35.384596162Z created by net/http.(*Server).Serve in goroutine 60

2024-08-12T09:33:35.384598662Z 	/usr/local/go/src/net/http/server.go:3285 +0x4b4
```

### CODE REVIEW

File: upload-server/pkg/redislocker/redislocker.go [hunk 1]
```diff
@@ -27,32 +28,70 @@ func WithLogger(logger *slog.Logger) LockerOption {
...
 	if err != nil {
 		return nil, err
 	}
 	client := redis.NewClient(connection)
-	if res := client.Ping(context.Background()); res.Err() != nil {
-		return nil, res.Err()
+
+	rsm := &RedSyncMutex{
+		conn: connection,
 	}
```
Comment (whytheplatypus) [inline upload-server/pkg/redislocker/redislocker.go:63]: with the default options this isn't needed.

File: upload-server/pkg/redislocker/redislocker.go [hunk 2]
```diff
@@ -27,32 +28,70 @@ func WithLogger(logger *slog.Logger) LockerOption {
...
+	for _, option := range defaultOpts {
 		option(locker)
 	}
-	//defaults
-	if locker.logger == nil {
-		locker.logger = slog.Default()
+
+	for _, option := range lockerOptions {
+		option(locker)
 	}
```
Comment (whytheplatypus) [inline upload-server/pkg/redislocker/redislocker.go:77]: Could assert that CreateMutex and logger aren't nil here after all the options run just to be extra defensive

**Review by whytheplatypus — CHANGES_REQUESTED** [review]

File: upload-server/pkg/redislocker/redislocker.go (hunk 1 above)
Comment (cfarmer-fearless) [reply upload-server/pkg/redislocker/redislocker.go]: removed

File: upload-server/pkg/redislocker/redislocker.go (hunk 2 above)
Comment (cfarmer-fearless) [reply upload-server/pkg/redislocker/redislocker.go:72]: done

**Comment by whytheplatypus** [comment]
can the `redislockerhealth` module now be removed?

**Review by whytheplatypus — APPROVED** [review]

**Pull request merged**